chore(windows): remove legacy core and flag

rc-swag opened the pull request:
Fixes: #5442 

Removing the legacy code that was replaced by the core. Also removing the feature flag for core integration.


# User Testing

# TEST_FULL_REGRESION_STUB
TODO: complete a full regression test on this.

rc-swag commented on windows/src/engine/keyman32/keyman32.cpp:
```diff
@@ -563,64 +563,55 @@ extern "C" BOOL  _declspec(dllexport) WINAPI Keyman_ForceKeyboard(PCSTR FileName
+  RefreshPreservedKeys(TRUE);
+  return TRUE;
+    // happy to use while(!done) pattern
+fail:
```
I realised there was previously a possible memory leak here in not deleting `_td_>lpActiveKeyboard` I am using a goto here. If we don't like that I could maybe use while(!done) {} instead

bharanidharanj commented:
## SUITE_BASELINE: Keyman for Windows Base Line Tests

### GROUP_WIN10: 

* **TEST_INSTALL (PASSED):** Tested with the attached Keyman build (17.0.87-alpha-test-8593) in Windows 10 OS and I was able to installed this build without having any problem. 
* **TEST_KEYBOARD_INSTALLATION_REMOTE (PASSED):** Able to install Khmer Angkor keyboard using Download Keyboard button in the Keyman Configuration dialog. 
* **TEST_INSTALL_PKG_DISK (PASSED):** Able to install sil_ipa.kmp file from a folder. But, I noticed that the previously added 'Khmer Angkor' keyboard is missing in the Keyman menu list in the task bar. 

   <img width="300" src="https://user-images.githubusercontent.com/19683143/233393583-3362b074-f0a8-4c02-bc13-b0f0e25727f5.PNG">

bharanidharanj commented:
## SUITE_BASELINE: Keyman for Windows Base Line Tests

### GROUP_WIN10: 

* **TEST_KEYBOARD_OUTPUT (FAILED):** I noticed that the installed 'khmer angkor' keyboard was missing the in the Keyman menu even though it appears on the Keyman Configuration dialog. Seems to be an issue.

   <img width="300" src="https://user-images.githubusercontent.com/19683143/233400274-a7c862f5-ab90-4548-ac4c-4d31ae826b6c.PNG">


* **TEST_ON_SCREEN_KEYBOARD (PASSED):** Verified that OSK displayed correctly for the selected keyboard. Able to type in the Notepad using OSK.

mcdurdin reviewed (approved):
LGTM. I wonder if there are other functions we can cleanup as a result of this (e.g. extended string management, etc). But those can come as separate work.

mcdurdin commented:
I think we can remove `LPKEYBOARD Keyboard;` from keymanengine.h, along with deleting legacy_kmx_memory.h altogether, yes? And that should filter out a bunch of other obsolete bits and pieces I think?

mcdurdin commented:
addins.cpp/addins.h should also be deleteable I think

rc-swag commented on windows/src/engine/keyman32/k32_globals.cpp:
```diff
@@ -385,9 +380,6 @@ void Globals::SetBaseKeyboardFlags(char *baseKeyboard, BOOL simulateAltGr, BOOL
 */
 BOOL Globals::InitSettings() {
-  /* Check for common core vs windows core */
-  f_CoreIntegration = Reg_GetDebugFlag(REGSZ_Flag_UseKeymanCore, TRUE);
```
ahh missed that thanks

bharanidharanj commented:
## SUITE_CAPSLOCK: 

### GROUP_WIN10: 

* **TEST_CAPSLOCK-1 (PASSED):** Verified that 'pass' result shown in the Notepad.
* **TEST_CAPSLOCK-2 (PASSED):** Verified that 'pass' result shown in the Notepad.
* **TEST_CAPSLOCK-3 (PASSED):** Verified that 'pass' result shown in the Notepad.
* **TEST_CAPSLOCK-4 (PASSED):** Verified that 'pass' result shown in the Notepad.
* **TEST_CAPSLOCK-5 (PASSED):** Verified that 'pass' result shown in the Notepad.
* **TEST_CAPSOFF-1 (PASSED):** Verified that 'ncaps_little_a' result shown in the Notepad.
* **TEST_CAPSOFF-2 (PASSED):** Verified that 'ncaps_little_a' result shown in the Notepad.
* **TEST_CAPSOFF-3 (PASSED):** Verified that 'ncaps_little_a' result shown in the Notepad.
* **TEST_CAPSOFF-4 (PASSED):** Verified that 'ncaps_little_a' result shown in the Notepad.
* **TEST_CAPSOFF-5 (PASSED):** Verified that 'ncaps_little_a' result shown in the Notepad.
* **TEST_CAPSONLY-1 (PASSED):** Verified that 'pass' result shown in the Notepad. 
* **TEST_CAPSONLY-2 (PASSED):** Verified that 'pass' result shown in the Notepad.
* **TEST_CAPSONLY-3 (PASSED):** Verified that 'pass' result shown in the Notepad.
* **TEST_CAPSONLY-4 (PASSED):** Verified that 'pass' result shown in the Notepad.
* **TEST_CAPSONLY-5 (PASSED):** Verified that nothing has output in the Notepad.

rc-swag commented on windows/src/engine/keyman32/kmhook_getmessage.cpp:
```diff
@@ -252,13 +252,9 @@ LRESULT _kmnGetMessageProc(int nCode, WPARAM wParam, LPARAM lParam)
 	*/
 
-	if(mp->message == WM_UNICHAR && Addin_ShouldProcessUnichar(mp->hwnd))
+	if(mp->message == WM_UNICHAR)
```
@mcdurdin I blindly removed addins - I don't know what they did. Should `WM_UNICHAR` be removed totally as it seems addins was the only function interested in the event. ?

mcdurdin replied:
I think `WM_UNICHAR` is very much dead at this point. But perhaps it should be removed entirely in a separate PR (and also remove Keyman_ForceKeyboard in a separate PR too) -- that way we have PR# to reference them for potential future questions.

rc-swag replied:
PR for removing `WM_UNICHAR` https://github.com/keymanapp/keyman/issues/8677

bharanidharanj commented:
## SUITE_TSF_APP: 

### GROUP_WIN10: 

* **TEST_DOUBLE_PROCESSING_FIREFOX (PASSED):** Verified that it shows the expected result abc

* **TEST_DOUBLE_PROCESSING_SEARCHBAR (PASSED):** Verified that it shows the expected result abc

* **TEST_DOUBLE_PROCESSING_NOTEPAD (PASSED):** Verified that it shows the expected result abcd

bharanidharanj commented:
## SUITE_IMX_KEYBOARDS: 

### GROUP_WIN10: 

* **TEST_IMSAMPLE_INPUT (PASSED):** Verified the expected result is try æ
* **TEST_IMSAMPLE_INPUT_CONT (PASSED):** Verified the expected result is æ###
* **TEST_IMSAMPLE_BACKSPACE (PASSED):** Verified the expected results is f
* **TEST_IMSAMPLE_INPUT_IMW (PASSED):** Verified the expected result is try æ
* **TEST_IMSAMPLE_INPUT_IMW_CONT (PASSED):** Verified the expected result is æ###
* **TEST_SIMPLIFIED_CHINESE_SINGLE (PASSED):** Verified the expected result is 喔
* **TEST_SIMPLIFIED_CHINESE_MULTIPLE (PASSED):** Verified the expected result is  汉字变换
* **TEST_SIMPLIFIED_CHINESE_BACKSPACE_1 (PASSED):** Verified the expected result is 汗
* **TEST_SIMPLIFIED_CHINESE_BACKSPACE_2 (PASSED):** Verified the expected result is 汉字

bharanidharanj commented:
## SUITE_STORE_AND_CONTEXT: 

### GROUP_WIN10: 

* **TEST_21_OPTION_STORE (PASSED):** Verified the expected result shown 'no foo.foo.no foo' in the Notepad.
* **TEST_OUTPUT_KEYSTROKE (PASSED):** Verified the expected result is 'abcd3'
* **TEST_OUTPUT_KEYSTROKE_INVALID_CONTEXT (PASSED):** Verified the expected result is 
abc
                                                                                                                                       3
* **TEST_DEADKEY_AND_CONTEXT (PASSED):** Verified the expected result is 'correct'

bharanidharanj commented:
## SUITE_BASELINE: Keyman for Windows Base Line Tests

### GROUP_WIN11: 

* **TEST_INSTALL (PASSED):** Tested with the attached Keyman build (17.0.87-alpha-test-8593) in Windows 11 OS (Virtual Machine) and I was able to installed this build without having any problem.
* **TEST_KEYBOARD_INSTALLATION_REMOTE (PASSED):** Able to install Khmer Angkor keyboard using Download Keyboard button in the Keyman Configuration dialog.
* **TEST_INSTALL_PKG_DISK (PASSED):** Able to install sil_ipa.kmp file from a folder. Verified that the keyboard appears on the menu.
* **TEST_KEYBOARD_OUTPUT (PASSED):** Able to change the keyboards and type it in the relevant language in the Notepad. 
* **TEST_ON_SCREEN_KEYBOARD (PASSED):** Verified that OSK displayed correctly for the selected keyboard. Able to type in the Notepad using OSK.

bharanidharanj commented:
## SUITE_CAPSLOCK: 

### GROUP_WIN11: 

* **TEST_CAPSLOCK-1 (PASSED):** Verified that 'pass' result shown in the Notepad.
* **TEST_CAPSLOCK-2 (PASSED):** Verified that 'pass' result shown in the Notepad.
* **TEST_CAPSLOCK-3 (PASSED):** Verified that 'pass' result shown in the Notepad.
* **TEST_CAPSLOCK-4 (PASSED):** Verified that 'pass' result shown in the Notepad.
* **TEST_CAPSLOCK-5 (PASSED):** Verified that 'pass' result shown in the Notepad.
* **TEST_CAPSOFF-1 (PASSED):** Verified that 'ncaps_little_a' result shown in the Notepad.
* **TEST_CAPSOFF-2 (PASSED):** Verified that 'ncaps_little_a' result shown in the Notepad.
* **TEST_CAPSOFF-3 (PASSED):** Verified that 'ncaps_little_a' result shown in the Notepad.
* **TEST_CAPSOFF-4 (PASSED):** Verified that 'ncaps_shift_A' result shown in the Notepad.
* **TEST_CAPSOFF-5 (PASSED):** Verified that 'ncaps_little_a' result shown in the Notepad.
* **TEST_CAPSONLY-1 (PASSED):** Verified that 'pass' result shown in the Notepad.
* **TEST_CAPSONLY-2 (PASSED):** Verified that 'pass' result shown in the Notepad.
* **TEST_CAPSONLY-3 (PASSED):** Verified that 'pass' result shown in the Notepad.
* **TEST_CAPSONLY-4 (PASSED):** Verified that 'pass' result shown in the Notepad.
* **TEST_CAPSONLY-5 (PASSED):** Verified that nothing has output in the Notepad.

bharanidharanj commented:
## SUITE_TSF_APP: 

### GROUP_WIN11: 

* **TEST_DOUBLE_PROCESSING_FIREFOX (PASSED):** Verified that it shows the expected result abc
* **TEST_DOUBLE_PROCESSING_SEARCHBAR (PASSED):** Verified that it shows the expected result abc
* **TEST_DOUBLE_PROCESSING_NOTEPAD (PASSED):** Verified that it shows the expected result abcd

bharanidharanj commented:
## SUITE_IMX_KEYBOARDS: 

### GROUP_WIN11: 

* **TEST_IMSAMPLE_INPUT (FAILED):** Installed the imsample keyboard in Windows 11 OS. Noticed that after pressing 1 or choosing the 1 from the Keyman Sample IM window is showing the expected output on the Notepad. Seems to be an issue. 

   <img width="300" src="https://user-images.githubusercontent.com/19683143/233644261-a7870559-f84b-407a-a0ee-5defecd02f19.PNG">

* **TEST_IMSAMPLE_INPUT_CONT (FAILED):** Not able to choose the letters from the Keyman Sample IM Window. 
* **TEST_IMSAMPLE_BACKSPACE (FAILED):** Not able to choose the letters from the Keyman Sample IM Window. 
* **TEST_IMSAMPLE_INPUT_IMW (FAILED):** Not able to choose the letters from the Keyman Sample IM Window. 
* **TEST_IMSAMPLE_INPUT_IMW_CONT (FAILED):** Not able to choose the letters from the Keyman Sample IM Window.

mcdurdin commented on windows/src/engine/keyman32/keyman32.cpp:
```diff
@@ -595,7 +591,10 @@ extern "C" BOOL  _declspec(dllexport) WINAPI Keyman_ForceKeyboard(PCSTR FileName
   }
 
-  ResetCapsLock();
+  // TODO: #5882 Add km_kbp_event to reset keyboard action sent to keyman core
+  // (only the core knows the caps rules such CAPS_ALWAYS_OFF)
+  // so it can then respond with a possible reset
```
Does this need to be an issue for resolution?

mcdurdin commented on windows/src/engine/keyman32/selectkeyboard.cpp:
```diff
@@ -101,7 +101,8 @@ BOOL SelectKeyboard(DWORD KeymanID)
         if (_td->app) _td->app->ResetContext();
-        ResetCapsLock();
+
+        // TODO: #5882 tell the core with km_kbp_event so it can reset the capslock state
```
Ditto question on issue tracking this

mcdurdin reviewed (approved):
Re-approving post addins code removal

mcdurdin commented:
Developer build failure can be ignored -- due to updated `build.sh pack` command for local build of npm packages; this will be available once merged.

rc-swag commented on windows/src/engine/keyman32/keyman32.cpp:
```diff
@@ -563,64 +563,55 @@ extern "C" BOOL  _declspec(dllexport) WINAPI Keyman_ForceKeyboard(PCSTR FileName
 	_td->lpActiveKeyboard->IMDLLs = NULL;
   _td->lpActiveKeyboard->KeyboardOptions = NULL;*/
```
See Issue #8646

rc-swag commented:
@bharanidharanj  I have reproduced the failures you have found and will move this back to draft to investigate.

rc-swag commented:
I made mistake when removing the addin.h/addin.c from the project. In the hook message for UNI_CHAR I should have simply done nothing. However, I moved to the text message. Regardless I have reverted the commit and created a separate PR just to remove addins for easier tracking. It can simply cherry-pick the commit but handle the message correctly.

rc-swag commented:
@keymanapp-test-bot retest SUITE_BASELINE GROUP_WIN10 TEST_KEYBOARD_OUTPUT

rc-swag commented:
In Investing the failed test for Windows 11 on `SUITE_IMX_KEYBOARDS` it appears that these tests also fail for Keyman 14 and 16. Issue #8697 has been created to address this issue.  
So the failing  `SUITE_IMX_KEYBOARDS` , `GROUP_WIN11` will be addressed in issue #8697

rc-swag commented on windows/src/engine/keyman32/K32_load.cpp:
```diff
@@ -114,46 +114,23 @@ BOOL LoadlpKeyboardCore(int i)
   LoadDLLs(&_td->lpKeyboards[i]);
 
-  LoadKeyboardOptions(&_td->lpKeyboards[i]);
+  LoadKeyboardOptionsREGCore(&_td->lpKeyboards[i], _td->lpKeyboards[i].lpCoreKeyboardState);
```
Good question. The `LoadlpKeyboardCore` was more about distinguishing it from the legacy equivalent function. The `LoadKeyboardOptionsREGCore ` is more about loading registry options to Core. However, it made me think and go over all the functions with Core.  I will create PR to clean this up.

rc-swag commented:
Since @mcdurdin last approved it 3 commits have been made and it is now ready for review again.   
In an attempt to keep this PR from getting harder to review, the following separate issues have been made. 
- [ ] https://github.com/keymanapp/keyman/issues/8681
- [ ] https://github.com/keymanapp/keyman/issues/8677
- [ ] https://github.com/keymanapp/keyman/issues/8703
- [ ] https://github.com/keymanapp/keyman/issues/8646

The failing `SUITE_IMX_KEYBOARDS` , `GROUP_WIN11` will be addressed in issue #8697. This is not a result of this PR but rather discovered in testing this PR.

bharanidharanj commented:
# Test Results

## SUITE_BASELINE: Keyman for Windows Base Line Tests

### GROUP_WIN10: 

* **TEST_KEYBOARD_OUTPUT (OPEN):** notes

> @bharanidharanj I have reproduced the failures you have found and will move this back to draft to investigate.

Okay, Thanks @rc-swag

mcdurdin commented on windows/src/engine/keyman32/keymanengine.h:
```diff
@@ -105,17 +103,12 @@ typedef struct tagKMSTATE
   WCHAR charCode;      // I4582
   BOOL windowunicode;  // I4287
   BOOL isDown;
   LPKEYBOARD lpkb;
```
Is this needed now?

mcdurdin reviewed (approved):
re-approving

ermshiperete commented on windows/src/engine/keyman32/K32_load.cpp:
```diff
@@ -75,17 +69,17 @@ BOOL GetKeyboardFileName(LPSTR kbname, LPSTR buf, int nbuf)
+BOOL LoadlpKeyboard(int i)
 {
-  SendDebugMessageFormat(0, sdmLoad, 0, "LoadlpKeyboardCore: Enter ---");
+  SendDebugMessageFormat(0, sdmLoad, 0, "LoadlpKeyboard: Enter ---");
```
You could use `__FUNCTION__` instead of hardcoding the function name:

```suggestion
  SendDebugMessageFormat(0, sdmLoad, 0, "%s: Enter ---", __FUNCTION__);
```

keyman-server commented:
Changes in this pull request will be available for download in [Keyman version 17.0.109-alpha](https://keyman.com/downloads/releases/alpha/17.0.109)
